docs/samples/diesel_small_config.md
```c
#define DASHBOARD_MESSAGE_MAX_LENGTH 18

#define DASHBOARD_ITEMS                            \
    X(FIRMWARE_ITEM, "GiUCAN " GIUCAN_VERSION)     \
    X(UPTIME_ITEM, "Uptime: %.0fmin")              \
    X(HP_ITEM, "Power: %.1fhp")                    \
    X(NM_ITEM, "Torque: %.0fnm")                   \
    X(DPF_STATUS_ITEM, "DPF: %s")                  \
    X(DPF_CLOG_ITEM, "DPF clog: %.0f%%")           \
    X(DPF_TEMP_ITEM, "DPF t: %.0f"                 \
                     "\xB0"                        \
                     "C")                          \
    X(DPF_REG_ITEM, "DPF ref: %.0f%%")             \
    X(DPF_DIST_ITEM, "DPF dist: %.0fkm")           \
    X(DPF_COUNT_ITEM, "DPF count: %.0f")           \
    X(DPF_MEAN_DIST_ITEM, "DPF mean: %.0fkm")      \
    X(DPF_MEAN_DURATION_ITEM, "DPF mean: %.0fmin") \
    X(BATTERY_V_ITEM, "Battery: %.1fV")            \
    X(BATTERY_A_ITEM, "Battery: %.2fA")            \
    X(BATTERY_P_ITEM, "Battery: %.0f%%")           \
    X(OIL_PRESS_ITEM, "Oil p: %.1fbar")            \
    X(OIL_QUALITY_ITEM, "Oil q: %.0f%%")           \
    X(OIL_TEMP_ITEM, "Oil t: %.0f"                 \
                     "\xB0"                        \
                     "C")                          \
    X(COOLANT_TEMP_ITEM, "Coolant t: %.0f"         \
                         "\xB0"                    \
                         "C")                      \
    X(AIR_IN_ITEM, "Air in t: %.0f"                \
                   "\xB0"                          \
                   "C")                            \
    X(GEAR_ITEM, "Gear: %c")                       \
    X(GEARBOX_TEMP_ITEM, "Gearbox t: %.0f"         \
                         "\xB0"                    \
                         "C")                      \
    X(STEERING_ITEM, "Steering: %.1f"              \
                     "\xB0")                       \
    X(TIRES_TEMP_FRONT_ITEM, "%.0f"                \
                             "\xB0"                \
                             "C ^ T ^ %.0f"        \
                             "\xB0"                \
                             "C")                  \
    X(TIRES_TEMP_REAR_ITEM, "%.0f"                 \
                            "\xB0"                 \
                            "C v T v %.0f"         \
                            "\xB0"                 \
                            "C")

#define CONVERTERS                                                             \
    X(DPF_STATUS_ITEM, const char *, dpf_status_as_string(value), bool, false) \
    X(GEAR_ITEM, char, ((unsigned char)value), bool, false)

#define EXTRACTION_FUNCTIONS                                                      \
    X(extractTempCommon, ((float)(((A(r) * 256) + B(r))) * 0.02f) - 40.0f)        \
    X(extractCarUptime, ((float)((A(r) * 256) + B(r)) / 4.0f))                    \
    X(extractHP, ((float)s->car.torque - 500) * (float)s->car.rpm * 0.000142378f) \
    X(extractNM, (float)s->car.torque - 500)                                      \
    X(extractDpfStatus, (float)s->car.dpf.regenMode)                              \
    X(extractDpfClog, ((float)((A(r) * 256) + B(r))) * 0.01525902f)               \
    X(extractDpfReg, ((float)((A(r) * 256) + B(r))) * 0.001525902f)               \
    X(extractDpfDist, ((float)((A(r) * 65536) + (B(r) * 256) + C(r))) * 0.1)      \
    X(extractDpfCount, (float)((A(r) * 256) + B(r)))                              \
    X(extractDpfMeanDist, (float)((A(r) * 256) + B(r)))                           \
    X(extractDpfMeanDuration, (float)((A(r) * 256) + B(r)) / 60.0f)               \
    X(extractBatteryVolt, (float)((A(r) * 256) + B(r)) * 0.0005f)                 \
    X(extractBatteryPerc, (float)s->car.battery.chargePercent)                    \
    X(extractBatteryAmpere, (float)s->car.battery.current)                        \
    X(extractOilPressure, s->car.oil.pressure)                                    \
    X(extractOilQuality, ((float)((A(r) * 256) + B(r))) * 0.001525902f)           \
    X(extractOilTemp, (float)s->car.oil.temperature)                              \
    X(extractGearboxTemp, (float)A(r) - 40.0f)                                    \
    X(extractGear, (float)s->car.gear)                                            \
    X(extractSteeringAngle, ((float)((((int8_t)A(r)) * 256) + B(r))) / 16.0f)

#define EXTRACTORS                                                                                                          \
    X(UPTIME_ITEM, true, true, 0x18DA10F1, 0x03221009, extractCarUptime, false, false, 0, 0, noop_extract)                  \
    X(HP_ITEM, true, false, 0, 0, extractHP, false, false, 0, 0, noop_extract)                                              \
    X(NM_ITEM, true, false, 0, 0, extractNM, false, false, 0, 0, noop_extract)                                              \
    X(DPF_STATUS_ITEM, true, false, 0, 0, extractDpfStatus, false, false, 0, 0, noop_extract)                               \
    X(DPF_CLOG_ITEM, true, true, 0x18DA10F1, 0x032218E4, extractDpfClog, false, false, 0, 0, noop_extract)                  \
    X(DPF_TEMP_ITEM, true, true, 0x18DA10F1, 0x032218DE, extractTempCommon, false, false, 0, 0, noop_extract)               \
    X(DPF_REG_ITEM, true, true, 0x18DA10F1, 0x0322380B, extractDpfReg, false, false, 0, 0, noop_extract)                    \
    X(DPF_DIST_ITEM, true, true, 0x18DA10F1, 0x03223807, extractDpfDist, false, false, 0, 0, noop_extract)                  \
    X(DPF_COUNT_ITEM, true, true, 0x18DA10F1, 0x032218A4, extractDpfCount, false, false, 0, 0, noop_extract)                \
    X(DPF_MEAN_DIST_ITEM, true, true, 0x18DA10F1, 0x03223809, extractDpfMeanDist, false, false, 0, 0, noop_extract)         \
    X(DPF_MEAN_DURATION_ITEM, true, true, 0x18DA10F1, 0x0322380A, extractDpfMeanDuration, false, false, 0, 0, noop_extract) \
    X(BATTERY_V_ITEM, true, true, 0x18DA10F1, 0x03221955, extractBatteryVolt, false, false, 0, 0, noop_extract)             \
    X(BATTERY_A_ITEM, true, false, 0, 0, extractBatteryAmpere, false, false, 0, 0, noop_extract)                            \
    X(BATTERY_P_ITEM, true, false, 0, 0, extractBatteryPerc, false, false, 0, 0, noop_extract)                              \
    X(OIL_PRESS_ITEM, true, false, 0, 0, extractOilPressure, false, false, 0, 0, noop_extract)                              \
    X(OIL_QUALITY_ITEM, true, true, 0x18DA10F1, 0x03223813, extractOilQuality, false, false, 0, 0, noop_extract)            \
    X(OIL_TEMP_ITEM, true, false, 0, 0, extractOilTemp, false, false, 0, 0, noop_extract)                                   \
    X(COOLANT_TEMP_ITEM, true, true, 0x18DA10F1, 0x03221003, extractTempCommon, false, false, 0, 0, noop_extract)           \
    X(AIR_IN_ITEM, true, true, 0x18DA10F1, 0x03221935, extractTempCommon, false, false, 0, 0, noop_extract)                 \
    X(GEAR_ITEM, true, false, 0, 0, extractGear, false, false, 0, 0, noop_extract)                                          \
    X(GEARBOX_TEMP_ITEM, true, true, 0x18DA18F1, 0x032204FE, extractGearboxTemp, false, false, 0, 0, noop_extract)          \
    X(STEERING_ITEM, true, true, 0x18DA2AF1, 0x0322083C, extractSteeringAngle, false, false, 0, 0, noop_extract)

#ifndef VALUES_REFRESH_MS
/* item_type, values_refresh_ms */
#define VALUES_REFRESH_MS            \
    X(FIRMWARE_ITEM, 60000)          \
    X(UPTIME_ITEM, 1000)             \
    X(DPF_STATUS_ITEM, 1000)         \
    X(DPF_CLOG_ITEM, 5000)           \
    X(DPF_TEMP_ITEM, 5000)           \
    X(DPF_REG_ITEM, 1000)            \
    X(DPF_DIST_ITEM, 60000)          \
    X(DPF_COUNT_ITEM, 60000)         \
    X(DPF_MEAN_DIST_ITEM, 60000)     \
    X(DPF_MEAN_DURATION_ITEM, 60000) \
    X(BATTERY_V_ITEM, 1000)          \
    X(BATTERY_A_ITEM, 1000)          \
    X(BATTERY_P_ITEM, 10000)         \
    X(OIL_QUALITY_ITEM, 60000)       \
    X(GEARBOX_TEMP_ITEM, 10000)      \
    X(TIRES_TEMP_FRONT_ITEM, 15000)  \
    X(TIRES_TEMP_REAR_ITEM, 15000)

#endif
```


docs/samples/gasoline_large_config.md
```c
#define DISABLE_DPF_REGEN_NOTIFICATION

#define DASHBOARD_ITEMS                         \
    X(FIRMWARE_ITEM, "GiUCAN " GIUCAN_VERSION)  \
    X(UPTIME_ITEM, "Uptime: %.0fmin/%.0fmin")   \
    X(HP_NM_ITEM, "Power: %.1fhp/%.0fnm")       \
    X(BATTERY_V_A_ITEM, "Battery: %.1fV/%.2fA") \
    X(BATTERY_P_ITEM, "Battery charge: %.0f%%") \
    X(OIL_TEMP_ITEM, "Oil temp.: %.0f"          \
                     "\xB0"                     \
                     "C")                       \
    X(COOLANT_TEMP_ITEM, "Coolant temp.: %.0f"  \
                         "\xB0"                 \
                         "C")                   \
    X(AIR_IN_ITEM, "Air in temp.: %.0f"         \
                   "\xB0"                       \
                   "C")                         \
    X(GEAR_ITEM, "Current gear: %c")            \
    X(GEARBOX_TEMP_ITEM, "Gearbox temp.: %.0f"  \
                         "\xB0"                 \
                         "C")                   \
    X(STEERING_ITEM, "Steering angle: %.1f"     \
                     "\xB0")
#define CONVERTERS \
    X(GEAR_ITEM, char, ((unsigned char)value), bool, false)

#define EXTRACTION_FUNCTIONS                                                      \
    X(extractTempCommon, ((float)(((A(r) * 256) + B(r))) * 0.02f) - 40.0f)        \
    X(extractCarUptime, ((float)((A(r) * 256) + B(r)) / 4.0f))                    \
    X(extractBoardUptime, (((float)s->board.now) / 60000.0f))                     \
    X(extractHP, ((float)s->car.torque - 500) * (float)s->car.rpm * 0.000142378f) \
    X(extractNM, (float)s->car.torque - 500)                                      \
    X(extractBatteryVolt, (float)((A(r) * 256) + B(r)) * 0.0005f)                 \
    X(extractBatteryPerc, (float)s->car.battery.chargePercent)                    \
    X(extractBatteryAmpere, (float)s->car.battery.current)                        \
    X(extractOilTemp, (float)s->car.oil.temperature)                              \
    X(extractGearboxTemp, (float)A(r) - 40.0f)                                    \
    X(extractGear, (float)s->car.gear)                                            \
    X(extractSteeringAngle, ((float)((((int8_t)A(r)) * 256) + B(r))) / 16.0f)

#define EXTRACTORS                                                                                                       \
    X(UPTIME_ITEM, true, true, 0x18DA10F1, 0x03221009, extractCarUptime, true, false, 0, 0, extractBoardUptime)          \
    X(HP_NM_ITEM, true, false, 0, 0, extractHP, true, false, 0, 0, extractNM)                                            \
    X(BATTERY_V_A_ITEM, true, true, 0x18DA10F1, 0x03221955, extractBatteryVolt, true, false, 0, 0, extractBatteryAmpere) \
    X(BATTERY_P_ITEM, true, false, 0, 0, extractBatteryPerc, false, false, 0, 0, noop_extract)                           \
    X(OIL_TEMP_ITEM, true, false, 0, 0, extractOilTemp, false, false, 0, 0, noop_extract)                                \
    X(COOLANT_TEMP_ITEM, true, true, 0x18DA10F1, 0x03221003, extractTempCommon, false, false, 0, 0, noop_extract)        \
    X(AIR_IN_ITEM, true, true, 0x18DA10F1, 0x03221935, extractTempCommon, false, false, 0, 0, noop_extract)              \
    X(GEAR_ITEM, true, false, 0, 0, extractGear, false, false, 0, 0, noop_extract)                                       \
    X(GEARBOX_TEMP_ITEM, true, true, 0x18DA18F1, 0x032204FE, extractGearboxTemp, false, false, 0, 0, noop_extract)       \
    X(STEERING_ITEM, true, true, 0x18DA2AF1, 0x0322083C, extractSteeringAngle, false, false, 0, 0, noop_extract)

#define VALUES_REFRESH_MS     \
    X(FIRMWARE_ITEM, 60000)   \
    X(UPTIME_ITEM, 1000)      \
    X(BATTERY_V_A_ITEM, 1000) \
    X(BATTERY_P_ITEM, 10000)  \
    X(GEARBOX_TEMP_ITEM, 10000)
```


inc/config.h
#ifndef __MAIN_H
#define __MAIN_H

#include "stm32f0xx_hal.h"

#define USB_FLASH_START_ADDRESS 0x8010000
#define TOTAL_USB_DEVICE_SIZE 65536

#ifdef INCLUDE_USER_CONFIG_H
#include "user_config.h"
#endif

#ifndef DISABLE_EXTERNAL_OSCILLATOR
#define ENABLE_EXTERNAL_OSCILLATOR
#endif

#ifndef GIUCAN_VERSION

#ifdef GIT_VERSION
#define GIUCAN_VERSION GIT_VERSION
#else
#define GIUCAN_VERSION "dev"
#endif

#endif

#ifdef SLCAN

#ifdef BHCAN
#error "Can't build SLCAN+BHCAN"
#endif

#ifdef C1CAN
#error "Can't build SLCAN+C1CAN"
#endif

#define LEDS_ON_CAN_RX
#define LEDS_ON_CAN_TX

#endif

#ifdef BHCAN

#define XCAN

#ifdef C1CAN
#error "Can't build BHCAN+C1CAN"
#endif

// 0x02=FM radio, 0x03=AM Radio, 0x05=Aux, 0x06=left USB, 0x07=Right USB,
// 0x08=Center USB, 0x09=Bluetooth, 0x12=phone connected, 0x13=phone disconnected,
// 0x15=call in progress, 0x17=call in wait, 0x18=call terminated, 0x11=clear display, ...
#ifndef DISPLAY_INFO_CODE
#define DISPLAY_INFO_CODE 0x08
#endif

// must be a multiple of 3; suggested value for 7 inch is 24, for 3.5 inch is 18
#ifndef DASHBOARD_MESSAGE_MAX_LENGTH
#define DASHBOARD_MESSAGE_MAX_LENGTH 24
#endif

#ifndef DISABLE_DASHBOARD_FORCED_REFRESH
#define DASHBOARD_FORCED_REFRESH
#ifndef DASHBOARD_FORCED_REFRESH_MS
#define DASHBOARD_FORCED_REFRESH_MS 1500
#endif
#endif

#ifndef DASHBOARD_FRAME_QUEUE_POLLING_INTERVAL_MS
#define DASHBOARD_FRAME_QUEUE_POLLING_INTERVAL_MS 29
#endif

#ifndef DISABLE_DPF_REGEN_NOTIFICATION
#define ENABLE_DPF_REGEN_NOTIFICATION

#ifndef DISABLE_DPF_REGEN_SOUND_NOTIFICATION
#define ENABLE_DPF_REGEN_SOUND_NOTIFICATION
#endif

#endif

#ifndef CAN_BITRATE
#define CAN_BITRATE CAN_BITRATE_125K
#endif

#endif

#ifdef C1CAN
#define XCAN

#ifndef DISABLE_DASHBOARD
#define ENABLE_DASHBOARD
#ifndef RES_LONG_PRESS_DURATION_MS
#define RES_LONG_PRESS_DURATION_MS 1000
#endif
#ifndef DASHBOARD_PAGE_SIZE
#define DASHBOARD_PAGE_SIZE 10
#endif
#endif

#ifndef DISABLE_SNS_AUTO_OFF
#define ENABLE_SNS_AUTO_OFF
#ifndef SNS_AUTO_OFF_DELAY_MS
#define SNS_AUTO_OFF_DELAY_MS 20000
#endif
#endif

#ifndef CAR_IS_ON_MIN_RPM
#define CAR_IS_ON_MIN_RPM 400
#endif

#ifndef ENABLE_DASHBOARD
#ifndef ENABLE_SNS_AUTO_OFF
#warning "You're building a C1CAN without any active feature :/"
#endif
#endif

#ifndef DEFAULT_VALUES_REFRESH_MS
#define DEFAULT_VALUES_REFRESH_MS 333
#endif

#ifndef VALUES_TIMEOUT_MS
#define VALUES_TIMEOUT_MS 60000
#endif

#ifndef STANDBY_DELAY_MS
#define STANDBY_DELAY_MS (VALUES_TIMEOUT_MS / 2)
#endif

#define STANDBY_MS (VALUES_TIMEOUT_MS+STANDBY_DELAY_MS)

#ifndef DISABLE_DPF_REGEN_NOTIFICATION
#define ENABLE_DPF_REGEN_NOTIFICATION

#ifndef DISABLE_DPF_REGEN_VISUAL_NOTIFICATION
#define ENABLE_DPF_REGEN_VISUAL_NOTIFICATION
#ifndef DPF_REGEN_VISUAL_NOTIFICATION_ITEM
#define DPF_REGEN_VISUAL_NOTIFICATION_ITEM DPF_STATUS_ITEM
#endif
#endif

#endif

#endif

#ifdef XCAN
#define PRINTF_INCLUDE_CONFIG_H
#ifndef DEBUG_MODE
#define ENABLE_USB_MASS_STORAGE
#ifdef C1CAN
#define ENABLE_RW_USB_MASS_STORAGE
#endif
#endif
#endif

#ifndef USART2_BAUD_RATE
#define USART2_BAUD_RATE 38400
#endif

#ifndef CAN_BITRATE
#define CAN_BITRATE CAN_BITRATE_500K
#endif

#if DEBUG_MODE
#ifdef ENABLE_DASHBOARD
#warning "DEBUG_MODE+ENABLE_DASHBOARD might lead to malfunction because of UART communication"
#endif
#ifdef BHCAN
#warning "DEBUG_MODE+BHCAN might lead to malfunction because of UART communication"
#endif
#endif

#endif /* __MAIN_H */
